### ISSUE TITLE
Fix for Big Sur libtool bug insufficient

### ISSUE THREAD

**Issue by jsquyres** [open-mpi/hwloc#447]
When I run `autogen.sh` on the hwloc HEAD (b7afa6c3881288fbe0b295d2ae709d9a640c756e) on MacOS Big Sur (11.1), I get an warning when it tries to patch `configure`:

```
$ ./autogen.sh
autoreconf: Entering directory `.'
autoreconf: configure.ac: not using Gettext
...etc...
autoreconf: Leaving directory `.'
Trying to patching configure for MacOS Big Sur libtool.m4 bug
patching file configure
Hunk #1 FAILED at 8846.
1 out of 1 hunk FAILED -- saving rejects to file configure.rej
WARNING: Couldn't apply Big Sur libtool.m4 bug fix.
```

This is apparently not an _error_; hwloc builds, but with lots of warnings like these:

```
  CC       topology.lo
warning: unknown warning option '-Wdiscarded-qualifiers'; did you mean
      '-Wignored-qualifiers'? [-Wunknown-warning-option]
warning: unknown warning option '-Wstack-usage=262144'; did you mean
      '-Wstack-exhausted'? [-Wunknown-warning-option]
warning: unknown warning option '-Wcast-function-type'; did you mean
      '-Wbad-function-cast'? [-Wunknown-warning-option]
```

I'm using Homebrew-installed tools:

* Autoconf 2.69
* Automake 1.16.3
* Libtool 2.4.6

I don't know if these are patched Autotools, and that's why the hwloc patch is failing...?

**Comment by bgoglin** [comment]
-Wstack-usage is likely unrelated? It gets enabled at configure time because gcc didn't error-out. Maybe we should add -Werror during this configure check?

Can you look at the generated configure script if it already contains the libtool fix?

**Comment by bgoglin** [comment]
Looks like it's clang warning about unknown options instead of erroring-out.

**Comment by jsquyres** [comment]
@bgoglin Ah, I see.  I didn't have time to investigate deeper before I posted, but your question was the right one: yes, the fix is already applied to `configure`, presumably by the Brew-patched-and-installed Libtool.

So the other warnings are secondary / different / unrelated to this.

Perhaps `autogen.sh` should check to see if the fix is already present before trying to patch...?  Perhaps something like this:

```
grep -A1 MACOSX_DEPLOYMENT_TARGET configure | grep powerpc
```

If the output is empty, the fix has not been applied.

**Comment by bgoglin** [comment]
I installed a patched libtool from Debian but the patch seems to be different. Here's the configure difference before we try to apply our patch:
```
@@ -8853,11 +8853,11 @@
       # to the OS version, if on x86, and 10.4, the deployment
       # target defaults to 10.4. Don't you love it?
       case ${MACOSX_DEPLOYMENT_TARGET-10.0},$host in
-	10.0,*86*-darwin8*|10.0,*-darwin[91]*)
+	10.0,*86*-darwin8*|10.0,*-darwin[912]*)
 	  _lt_dar_allow_undefined='$wl-undefined ${wl}dynamic_lookup' ;;
 	10.[012][,.]*)
 	  _lt_dar_allow_undefined='$wl-flat_namespace $wl-undefined ${wl}suppress' ;;
-	10.*)
+	10.*|11.*)
 	  _lt_dar_allow_undefined='$wl-undefined ${wl}dynamic_lookup' ;;
       esac
     ;;
```
It looks like it's a different patch but it seems considered less future-proof. I am going to grep for both. Can you test bgoglin/master?

@amckinstry I couldn't find the relevant commits on salsa, but can you comment on why you applied this patch instead of the other one from https://debbugs.gnu.org/cgi/bugreport.cgi?bug=44605 ?

**Comment by jsquyres** [comment]
@bgoglin bgoglin/master mostly works -- it has an amusing output glitch:

```
-n Checking whether configure needs patching for MacOS Big Sur libtool.m4 bug... 
no
```

I think it's because the script starts with `:` rather than `~#!/something`, meaning it uses the `echo` that is built in to the Bourne shell, and therefore doesn't grok `-n`.

Perhaps this would be better?

```patch
diff --git a/autogen.sh b/autogen.sh
index 96466acfb..d47256257 100755
--- a/autogen.sh
+++ b/autogen.sh
@@ -1,4 +1,4 @@
-:
+#!/usr/bin/env bash
 
 srcdir=$(dirname "$0")
 if test "x$srcdir" != x; then
```

**Comment by bgoglin** [comment]
Thanks, I'll fix it and push to master and 2.4.1rc2
